crucible-mir/src/Mir/Trans.hs

```diff
@@ -1618,6 +1618,46 @@ evalPlaceProj _ pl (M.Downcast _idx) = return pl
 -- Subtype is a no-op, as it is only present in the MIR to making subtyping
 -- explicit during optimizations and codegen.
 evalPlaceProj _ pl (M.Subtype _ty) = return pl
+-- Subslicing is defined on slices and arrays. See the haddock for `Subslice`
+-- for details on the semantics of `fromIndex` and `toIndex`.
+evalPlaceProj ty (MirPlace tpr ref meta) (M.Subslice fromIndex toIndex fromEnd) =
+  case (ty, ref, tpr, meta) of
+    (M.TySlice _elemTy, headRef, elemTpr, SliceMeta len) ->
+      let lastIndex = mkLastIndex len
+          newLen = R.App (lastIndex `usizeSub` firstIndex)
+          newMeta = SliceMeta newLen
+      in MirPlace elemTpr <$> mirRef_offset headRef firstIndex <*> pure newMeta
+
+    -- TODO: Crucible's representation of `MirReference`s to arrays is not
+    -- expressive enough to implement this case properly. A Crucible reference
+    -- to an array is a reference to the entire array, and cannot have
+    -- `mirRef_offset` applied to it - unlike slices, a Crucible reference to
+    -- which is represented as a reference to its first element.
+
+    -- A couple possible ways past this:
+    -- - Make Crucible array references actually point to the first element of
+    --   the array, mimicking Crucible's approach to slices and Rust itself's
+    --   approach to arrays. Then, they'd be amenable to `mirRef_offset`.
+    -- - Implement a new reference path extension, something like
+    --   "MirRef_Subarray". Such a path extension would behave somewhat like
+    --   `MirRef_Index`, but instead of offering read/write access to a single
+    --   element of the parent array, it would offer read/write access to a
+    --   subarray of elements of the parent array.
+    --
+    -- NB: after implementing this, update `Mir.Mir.typeOfProj`, which currently
+    -- declares that `Subslice` unconditionally yields a slice. This is
+    -- incorrect; `Subslice` applied to an array will yield an array.
```

Comment:
I think it would be helpful to turn all of this context into a `crucible` issue, then link back to the issue number here.

Reply:
Done - see #1494.